contracts/VaultV3.vy

```diff
@@ -354,121 +379,183 @@ def _issueSharesForAmount(amount: uint256, recipient: address) -> uint256:
-    if _sender != _owner: 
-      self._spendAllowance(_owner, _sender, _shares)
+    if _sender != _owner:
+        self._spendAllowance(_owner, _sender, _shares)
 
     shares: uint256 = _shares
     sharesBalance: uint256 = self.balanceOf[_owner]
 
-    if _shares == MAX_UINT256:
+    if shares == MAX_UINT256:
         shares = sharesBalance
 
     # TODO: is this needed? will revert in burn call
     assert sharesBalance >= shares, "insufficient shares to withdraw"
     assert shares > 0, "no shares to withdraw"
 
-    assets: uint256 = self._convertToAssets(_shares)
+    assets: uint256 = self._convertToAssets(shares)
 
-    # TODO: withdraw from strategies
-    
-    # TODO: gas savings: totalIdle should be cached if used above
-    assert self.totalIdle >= assets, "insufficient total idle"
+    if assets > self.totalIdle:
+        # load to memory to save gas
+        currTotalIdle: uint256 = self.totalIdle
+        currTotalDebt: uint256 = self.totalDebt
 
-    self._burnShares(_shares, _owner)
+        # withdraw from strategies if insufficient total idle
+        assetsNeeded: uint256 = assets - currTotalIdle
+        assetsToWithdraw: uint256 = 0
+        for strategy in _strategies:
+            assert self.strategies[strategy].activation != 0, "inactive strategy"
 
+            assetsToWithdraw = min(assetsNeeded, IStrategy(strategy).withdrawable())
+            # continue if nothing to withdraw
+            if assetsToWithdraw == 0:
+                continue
+
+            IStrategy(strategy).freeFunds(assetsToWithdraw)
+            ASSET.transferFrom(strategy, self, assetsToWithdraw)
+            currTotalIdle += assetsToWithdraw
+            currTotalDebt -= assetsToWithdraw
+            self.strategies[strategy].currentDebt -= assetsToWithdraw
+
+            # break if we have enough total idle
+            if assets <= currTotalIdle:
+                break
+
+            assetsNeeded -= assetsToWithdraw
+
+        # if we exhaust the queue and still have insufficient total idle, revert
+        assert currTotalIdle >= assets, "insufficient total idle"
+        # commit memory to storage
+        self.totalIdle = currTotalIdle
+        self.totalDebt = currTotalDebt
+
+    self._burnShares(shares, _owner)
     self.totalIdle -= assets
     self.erc20_safe_transfer(ASSET.address, _receiver, assets)
 
-    log Withdraw(_sender, _receiver, _owner, assets, _shares)
+    log Withdraw(_sender, _receiver, _owner, assets, shares)
 
     return assets
 
 
+@view
+@internal
+def _maxDeposit(receiver: address) -> uint256:
+    total_assets: uint256 = self._totalAssets()
+    deposit_limit: uint256 = self.depositLimit
+    if (total_assets >= deposit_limit):
+        return 0
+    return deposit_limit - total_assets
+
+
+@view
+@internal
+def _maxRedeem(owner: address) -> uint256:
+    return min(self.balanceOf[owner], self.totalIdle)
```

Comment:
self.balanceOf[owner] will return balance of shares held by owner, self.totalIdle will return assets

we should replace `self.totalIdle` with `self._convertToShares(self.totalIdle)`